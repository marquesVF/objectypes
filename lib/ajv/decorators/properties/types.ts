export type SharedOptions = {
  nullable?: boolean
}

export type NumberOptions = SharedOptions & {
  maximum?: number
  minimum?: number
  exclusiveMaximum?: number
  exclusiveMinimum?: number
}

export type NumberPropertyOptions = {
  expectedType: 'number'
  options?: NumberOptions
}
