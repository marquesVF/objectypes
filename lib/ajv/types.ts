export type ClassConstructor<T> = new () => T
export type LazyClassConstructor<T> = () => ClassConstructor<T>
