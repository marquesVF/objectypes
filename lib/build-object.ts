/* eslint-disable complexity */
import { path } from 'ramda'

import { Metadata } from './core/metadata'
import { Hashable, ClassConstructor } from './types'

export function buildObject<T>(
    targetKlass: ClassConstructor<Hashable & T>,
    jsonObj: Hashable
): T {
    const targetObj = new targetKlass()
    const metadataStorage = Metadata.getInstance()
    const properties = metadataStorage.findProperties(targetKlass)
    const transformations = metadataStorage
        .findTransformations(targetKlass, 'build')
    const reductions = metadataStorage.findReductions(targetKlass)

    if (properties) {
        for (const property of properties) {
            const { propertyKey, name, type, nullable, target } = property
            const objPropName = name ?? propertyKey

            if (reductions) {
                const reductionMetada = reductions
                        ?.find(metadata => metadata.propertyKey === propertyKey)
                if (reductionMetada) {
                    const value = reductionMetada.reducer.reduce(jsonObj)

                    Reflect.set(targetObj, propertyKey, value)
                    continue
                }
            }

            let value = path<any>(objPropName.split('.'), jsonObj) !== undefined
                ?  path<any>(objPropName.split('.'), jsonObj)
                : path<any>([propertyKey], jsonObj)

            if (value === undefined && !nullable) {
                throw new Error(
                    // eslint-disable-next-line max-len
                    `Property '${objPropName}' is missing. Couldn't build ${targetKlass.name} object.`
                )
            }

            const expectedType = Reflect
                .getMetadata('design:type', target, propertyKey)
                .name

            try {
                value = castValue(expectedType, value)
            } catch (err) {
                // eslint-disable-next-line max-len
                throw new Error(`Property ${objPropName} type is not assignable to ${expectedType}. Found ${value}`)
            }

            if (value !== undefined) {
                const transformMetadata = transformations
                    ?.find(metadata => metadata.propertyKey === propertyKey)
                if (transformMetadata) {
                    // TODO improve error handling since it may raise errors in runtine
                    value = transformMetadata.transformer.transform(value)
                }
            }

            if (type && value !== undefined) {
                const nestedValue = Array.isArray(value)
                    ? value.map(val => buildObject(type, val))
                    : buildObject(type, value)

                Reflect.set(targetObj, propertyKey, nestedValue)
            } else {
                Reflect.set(targetObj, propertyKey, value)
            }
        }
    }

    return targetObj
}

function castValue(expectedType: string, value?: any): any {
    if (value === undefined) {
        return value
    }

    if (expectedType === 'Number') {
        if (isNaN(value)) {
            throw new Error()
        }

        return Number(value)
    }
    if (expectedType === 'Date') {
        if (!Date.parse(value)) {
            throw new Error()
        }

        return new Date(value)
    }

    return value
}
