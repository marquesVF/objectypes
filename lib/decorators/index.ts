export { Property } from './property'
export { MapProperty } from './map-property'
export { BuildTransformation } from './build-transformation'
export { ExtractTransformation } from './extract-transformation'
