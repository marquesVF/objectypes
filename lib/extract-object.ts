import { assocPath } from 'ramda'

import { findClassPropertiesMetadata } from './core/metadata/property'
import { findClassTransformationMetadata } from './core/metadata/transformation'
import { Hashable, ClassConstructor, PropertyMetadata } from './types'
import { ExtractOptions } from './types/extract-options'

export function extractObject<T>(
  obj: Hashable & T,
  objClass: ClassConstructor<T>,
  options?: ExtractOptions
): object {
  let resultingObject: Hashable = {}
  const propertyMetadatas = findClassPropertiesMetadata(
    objClass,
    options?.namedOnly
  )

  if (propertyMetadatas) {
    for (const propertyMetadata of propertyMetadatas) {
      const { name, propertyKey, type } = propertyMetadata
      const value = obj[propertyKey]

      if (value === undefined) {
        continue
      }

      const transformedValue = applyTransformationsToObject(
        objClass,
        propertyMetadata,
        value
      )
      const finalValue = processNestedValue(transformedValue, type)
      const resultingProperty = name ?? propertyKey

      resultingObject = assocPath(
        resultingProperty.split('.'),
        finalValue,
        resultingObject
      )
    }
  }

  return resultingObject
}

function applyTransformationsToObject<T>(
  objectClass: ClassConstructor<T>,
  propertyMetadata: PropertyMetadata,
  value?: any
) {
  const { propertyKey } = propertyMetadata
  const transformationMetadatas = findClassTransformationMetadata(
    objectClass,
    'extract'
  )
  const transformationMetadata = transformationMetadatas?.find(
    metadata => metadata.propertyKey === propertyKey
  )

  if (!transformationMetadata) {
    return value
  }

  return transformationMetadata.transformer.transform(value)
}

function processNestedValue(value: any, type?: any) {
  if (!type) {
    return value
  }

  return Array.isArray(value)
    ? (value = value.map(val => extractObject(val, type)))
    : extractObject(value, type)
}
