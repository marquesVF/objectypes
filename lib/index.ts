export * from './types/property-options'
export * from './decorators'
export { extractObject } from './extract-object'
export { mapObject } from './map-object'
export { buildObject } from './build-object'
