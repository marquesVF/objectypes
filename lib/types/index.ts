export * from './class-constructor'
export * from './hashable'
export * from './map-property-metadata'
export * from './object-handler-validation'
export * from './property-metadata'
export * from './transform-function'
