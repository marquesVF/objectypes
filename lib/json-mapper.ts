/* eslint-disable no-useless-constructor */
import { ClassConstructor } from './types/class-constructor'
import { Hashable } from './types/hashable'
import { isValid } from './validate'
import { buildObject } from './build-object'
import { extractObject } from './extract-object'

export class JsonMapper<T> {

    readonly errors: string[] = []

    constructor(
        private readonly klass: ClassConstructor<T>
    ) { }

    validate(obj: Hashable): boolean {
        const { valid, errors } = isValid(this.klass, obj)

        if (errors) {
            this.errors.push(...errors)
        }

        return valid
    }

    validationErrorSummary(separator?: string): string {
        const actualSeparator = separator ?? ', '

        return this.errors.join(actualSeparator)
    }

    build(jsonObj: object): T {
        return buildObject(this.klass, jsonObj)
    }

    extract(obj: T): object {
        return extractObject(obj, this.klass)
    }

}
