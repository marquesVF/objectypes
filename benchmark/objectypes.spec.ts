import { buildObject } from '../lib'
import data from './data.json'
import { FilmResults } from './types/FilmResults'

describe('buildObject', () => {
  it('FilmResults object matches snapshot', () => {
    const filmResults = buildObject(FilmResults, data)

    expect(filmResults).toMatchSnapshot()
  })
})
