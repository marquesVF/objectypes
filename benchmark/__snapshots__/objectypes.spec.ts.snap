// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildObject FilmResults object matches snapshot 1`] = `
FilmResults {
  "page": 1,
  "results": [
    FilmResult {
      "adult": false,
      "id": 164558,
      "overview": "Go behind the scenes during One Directions sell out "Take Me Home" tour and experience life on the road.",
      "releaseDate": 2013-08-30T00:00:00.000Z,
      "title": "One Direction: This Is Us",
      "voteCount": 55,
    },
    FilmResult {
      "adult": false,
      "id": 654,
      "overview": "",
      "releaseDate": 1954-06-22T00:00:00.000Z,
      "title": "On the Waterfront",
      "voteCount": 51,
    },
    FilmResult {
      "adult": false,
      "id": 10911,
      "overview": "A team of allied saboteurs are assigned an impossible mission: infiltrate an impregnable Nazi-held island and destroy the two enormous long-range field guns that prevent the rescue of 2,000 trapped British soldiers.",
      "releaseDate": 1961-04-27T00:00:00.000Z,
      "title": "The Guns of Navarone",
      "voteCount": 50,
    },
    FilmResult {
      "adult": false,
      "id": 93,
      "overview": "The defendant In a murder trial says that he suffered temporary insanity after the victim had raped his wife. What is the truth, is his attorney being played, and will he win his case?",
      "releaseDate": 1959-07-01T00:00:00.000Z,
      "title": "Anatomy of a Murder",
      "voteCount": 37,
    },
    FilmResult {
      "adult": false,
      "id": 3078,
      "overview": "Ellie Andrews has just tied the knot with society aviator King Westley when she is whisked away to her father's yacht and out of King's clutches. Ellie jumps ship and eventually winds up on a bus headed back to her husband. Reluctantly she must accept the help of out-of- work reporter Peter Warne. Actually, Warne doesn't give her any choice: either she sticks with him until he gets her back to her husband, or he'll blow the whistle on Ellie to her father. Either way, Peter gets what he wants... a really juicy newspaper story!",
      "releaseDate": 1934-02-22T00:00:00.000Z,
      "title": "It Happened One Night",
      "voteCount": 48,
    },
    FilmResult {
      "adult": false,
      "id": 947,
      "overview": "Lawrence of Arabia is the classic film from David Lean starring Peter O’Toole and based on the autobiography from Thomas Edward Lawrence who during the first World War was on assignment by the British Empire in Arabia. The film would become a cult classic and is known today as a masterpiece.",
      "releaseDate": 1962-12-10T00:00:00.000Z,
      "title": "Lawrence of Arabia",
      "voteCount": 269,
    },
    FilmResult {
      "adult": false,
      "id": 3085,
      "overview": "Hildy Johnson has divorced Walter Burns and visits his office to tell him that she is engaged to another man and that they are going to get married the day after. Walter Burns can't let that happen and frames the other man, Bruce Baldwin, for a lot of stuff getting him into trouble all the time, while he tries to steer Hildy back into her old job as his employee (editor of his newspaper).",
      "releaseDate": 1940-01-11T00:00:00.000Z,
      "title": "His Girl Friday",
      "voteCount": 20,
    },
    FilmResult {
      "adult": false,
      "id": 14580,
      "overview": "Tough cop Dave Bannion takes on a politically powerful crime syndicate.",
      "releaseDate": 1953-10-14T00:00:00.000Z,
      "title": "The Big Heat",
      "voteCount": 10,
    },
    FilmResult {
      "adult": false,
      "id": 4584,
      "overview": "Rich Mr. Dashwood dies, leaving his second wife and her daughters poor by the rules of inheritance. Two daughters are the titular opposites.",
      "releaseDate": 1995-12-13T00:00:00.000Z,
      "title": "Sense and Sensibility",
      "voteCount": 82,
    },
    FilmResult {
      "adult": false,
      "id": 12102,
      "overview": "Ted Kramer is a career man for whom his work comes before his family. His wife Joanna cannot take this anymore, so she decides to leave him. Ted is now faced with the tasks of housekeeping and taking care of himself and their young son Billy. When he has learned to adjust his life to these new responsibilities, Joanna resurfaces and wants Billy back. Ted however refuses to give him up, so they go to court to fight for the custody of their son.",
      "releaseDate": 1979-12-19T00:00:00.000Z,
      "title": "Kramer vs. Kramer",
      "voteCount": 68,
    },
    FilmResult {
      "adult": false,
      "id": 1245,
      "overview": "A rule bound head butler's world of manners and decorum in the household he maintains is tested by the arrival of a housekeeper who falls in love with him in post-WWI Britain. The possibility of romance and his master's cultivation of ties with the Nazi cause challenge his carefully maintained veneer of servitude.",
      "releaseDate": 1993-11-12T00:00:00.000Z,
      "title": "The Remains of the Day",
      "voteCount": 37,
    },
    FilmResult {
      "adult": false,
      "id": 10436,
      "overview": "Tale of 19th century New York high society in which a young lawyer falls in love with a woman separated from her husband, while he is engaged to the woman's cousin.",
      "releaseDate": 1993-09-17T00:00:00.000Z,
      "title": "The Age of Innocence",
      "voteCount": 26,
    },
    FilmResult {
      "adult": false,
      "id": 9267,
      "overview": "And Now for Something Completely Different is a film spin-off from the television comedy series Monty Python's Flying Circus featuring favourite sketches from the first two seasons.",
      "releaseDate": 1971-09-28T00:00:00.000Z,
      "title": "And Now for Something Completely Different",
      "voteCount": 25,
    },
    FilmResult {
      "adult": false,
      "id": 137,
      "overview": "A narcissistic TV weatherman, along with his attractive-but-distant producer and mawkish cameraman, is sent to report on Groundhog Day in the small town of Punxsutawney, where he finds himself repeating the same day over and over.",
      "releaseDate": 1993-02-11T00:00:00.000Z,
      "title": "Groundhog Day",
      "voteCount": 549,
    },
    FilmResult {
      "adult": false,
      "id": 24807,
      "overview": "Longfellow Deeds lives in a small town, leading a small town kind of life. When a relative dies and leaves Deeds a fortune, Longfellow moves to the big city where he becomes an instant target for everyone. Deeds outwits them all until Babe Bennett comes along. When small-town boy meets big-city girl anything can, and does, happen.",
      "releaseDate": 1936-04-12T00:00:00.000Z,
      "title": "Mr. Deeds Goes to Town",
      "voteCount": 11,
    },
    FilmResult {
      "adult": false,
      "id": 293,
      "overview": "A River Runs Through is a cinematographically stunning true story of Norman Maclean. The story follows Norman and his brother Paul through the experiences of life and growing up, and how their love of fly fishing keeps them together despite varying life circumstances in the untamed west of Montana in the 1920's.",
      "releaseDate": 1992-10-09T00:00:00.000Z,
      "title": "A River Runs Through It",
      "voteCount": 69,
    },
    FilmResult {
      "adult": false,
      "id": 51052,
      "overview": "This Christmas movie highlights the technological advances of operations at the North Pole, revealing how Santa and his vast army of highly trained elves produce gifts and distribute them around the world in one night. However, every operation has a margin of error… When one of 600 million children to receive a gift from Santa on Christmas Eve is missed, it is deemed "acceptable" to all but one, Arthur. Arthur Claus is Santa’s misfit son who executes an unauthorized rookie mission to get the last present half way around the globe before dawn on Christmas morning.",
      "releaseDate": 2011-11-23T00:00:00.000Z,
      "title": "Arthur Christmas",
      "voteCount": 102,
    },
    FilmResult {
      "adult": false,
      "id": 13920,
      "overview": "Football coach Harold Jones befriends Radio, a mentally-challenged man who becomes a student at T.L. Hanna High School in Anderson, South Carolina. Their friendship extends over several decades, where Radio transforms from a shy, tormented man into an inspiration to his community.",
      "releaseDate": 2003-10-24T00:00:00.000Z,
      "title": "Radio",
      "voteCount": 36,
    },
    FilmResult {
      "adult": false,
      "id": 11076,
      "overview": "Amy is only 13 years old when her mother is killed. She goes to Canada to live with her father, an eccentric inventor whom she barely knows. Amy is miserable in her new life... until she discovers a nest of goose eggs that were abandoned when a local forest is torn down. The eggs hatch and Amy becomes "Mama Goose". When Winter comes, Amy, and her dad must find a way to lead the birds South...",
      "releaseDate": 1996-09-13T00:00:00.000Z,
      "title": "Fly Away Home",
      "voteCount": 13,
    },
    FilmResult {
      "adult": false,
      "id": 9587,
      "overview": "With their father away as a chaplain in the Civil War, Jo, Meg, Beth and Amy grow up with their mother in somewhat reduced circumstances. They are a close family who inevitably have their squabbles and tragedies. But the bond holds even when, later, male friends start to become a part of the household.",
      "releaseDate": 1994-12-21T00:00:00.000Z,
      "title": "Little Women",
      "voteCount": 50,
    },
  ],
  "totalPages": 4,
  "totalResults": 61,
}
`;
